transaction-pool/src/transactions.rs

```diff
@@ -0,0 +1,219 @@
+			let index = index + 1;
+			if index < self.scores.len() {
+				Some((self.scores[index].clone(), self.transactions[index].clone()))
+			} else {
+				None
+			}
+		})
+	}
+
+	fn push_cheapest_transaction(&mut self, tx: T, scoring: &S, max_count: usize) -> AddResult<T> {
+		let index = self.transactions.len();
+		if index == max_count {
+			AddResult::TooCheapToEnter(tx)
+		} else {
+			let shared = Arc::new(tx);
+			self.transactions.push(shared.clone());
+			self.scores.push(Default::default());
+			scoring.update_scores(&self.transactions, &mut self.scores, scoring::Change::InsertedAt(index));
+
+			AddResult::Ok(shared)
+		}
+	}
+
+	pub fn add(&mut self, tx: T, scoring: &S, max_count: usize) -> AddResult<T> {
+		let index = match self.transactions.binary_search_by(|old| scoring.compare(old, &tx)) {
+			Ok(index) => index,
+			Err(index) => index,
+		};
+
+		// Insert at the end.
+		if index == self.transactions.len() {
+			return self.push_cheapest_transaction(tx, scoring, max_count)
+		}
+
+		// Decide if the transaction should replace some other.
+		match scoring.choose(&self.transactions[index], &tx) {
+			// New transaction should be rejected
+			scoring::Choice::RejectNew => AddResult::TooCheap {
+				old: self.transactions[index].clone(),
+				new: tx,
+			},
+			// New transaction should be kept along with old ones.
+			scoring::Choice::InsertNew => {
+				let new = Arc::new(tx);
+
+				self.transactions.insert(index, new.clone());
+				self.scores.insert(index, Default::default());
+				scoring.update_scores(&self.transactions, &mut self.scores, scoring::Change::InsertedAt(index));
+
+				if self.transactions.len() > max_count {
+					let old = self.transactions.pop().expect("len is non-zero");
+					self.scores.pop();
+					scoring.update_scores(&self.transactions, &mut self.scores, scoring::Change::RemovedAt(self.transactions.len()));
+
+					AddResult::PushedOut {
+						old,
+						new,
+					}
+				} else {
+					AddResult::Ok(new)
+				}
+			},
+			// New transaction is replacing some other transaction already in the queue.
+			scoring::Choice::ReplaceOld => {
+				let new = Arc::new(tx);
+				let old = mem::replace(&mut self.transactions[index], new.clone());
+				scoring.update_scores(&self.transactions, &mut self.scores, scoring::Change::ReplacedAt(index));
+
+				AddResult::Replaced {
+					old,
+					new,
+				}
+			},
+		}
+	}
+
+	pub fn remove(&mut self, tx: &T, scoring: &S) -> bool {
+		let index = match self.transactions.binary_search_by(|old| scoring.compare(old, tx)) {
+			Ok(index) => index,
+			Err(_) => {
+				warn!("Attempting to remove non-existent transaction {:?}", tx);
+				return false;
+			},
+		};
+
+		self.transactions.remove(index);
+		self.scores.remove(index);
+		// Update scoring
+		scoring.update_scores(&self.transactions, &mut self.scores, scoring::Change::RemovedAt(index));
+		return true;
+	}
+
+	pub fn cull<R: Ready<T>>(&mut self, ready: &mut R, scoring: &S) -> SmallVec<[Arc<T>; PER_SENDER]> {
```

Comment:
Something makes me think, that this method may be refactored using iterators instead of `reverse`s.

Reply:
Since we use `SmallVec` it doesn't have methods that allow easily to consume the content.
Current method doesn't require allocation, but is indeed `O(n)`, since number of transactions per sender is limited I don't expect it to be a problem.

With `Vec` we could use `split_off`, although instead of `Vec` it would be more wise to use `VecDeque`, but that would require more allocations - might be worth to consider though after some initial tests.